include/glow/Graph/TensorLayout.h

```diff
@@ -0,0 +1,194 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_GRAPH_TENSORLAYOUT_H
+#define GLOW_GRAPH_TENSORLAYOUT_H
+
+#include <array>
+#include <memory>
+#include <string>
+
+#include "glow/Graph/Nodes.h"
+#include "glow/Support/Error.h"
+
+namespace glow {
+
+/// Layout requirements's Singleton.
+template <typename T> class TensorLayoutSingleton {
+public:
+  /// This is how the verifier, Backend and post-loading canonicalizer can
+  /// access layout constraints.
+  static T &getInstance() {
+    // The Ctor will only be called once.
+    static const std::unique_ptr<T> instance{new T{token_{}}};
+    return *instance;
+  }
+
+protected:
+  /// Allow the base class to call any subclass's constructor.
+  struct token_ {};
+
+  /// Default Ctor.
+  TensorLayoutSingleton() {}
+
+  /// Dtor.
+  virtual ~TensorLayoutSingleton() {}
+
+private:
+  /// Delete copy constructor.
+  TensorLayoutSingleton(const TensorLayoutSingleton &) = delete;
+
+  /// Delete move constructor.
+  TensorLayoutSingleton(TensorLayoutSingleton &&) = delete;
+
+  /// Delete copy assignment.
+  TensorLayoutSingleton &operator=(const TensorLayoutSingleton &) = delete;
+
+  /// Delete move assignment.
+  TensorLayoutSingleton &operator=(TensorLayoutSingleton &&) = delete;
+};
+
+/// TensorLayoutDescription - optional helper class for parsing string-based
+/// layout.
+class TensorLayoutDescription {
+  /// Tensor dimensions descriptions for all dimensions.
+  std::string dims_[max_tensor_dimensions];
+  /// The serialization of the layout.
+  std::string serializedLayout_;
+  /// Expected number of dimensions.
+  size_t numDims_;
+
+public:
+  virtual ~TensorLayoutDescription() = default;
+  /// Constructs this helper class from a serialized string representation.
+  TensorLayoutDescription(std::string text);
```

Comment:
Do you modify `text`? If not, should it be `const std::string &` or `StringRef`? Also, the name `text` is not very descriptive. Maybe `layout` or `layoutStr` or something like that?